crates/uv-toolchain/src/discovery.rs

```diff
@@ -452,185 +452,182 @@ fn should_stop_discovery(err: &Error) -> bool {
+                    }
+                })
+                .map(|result| result.map(Toolchain::from_tuple).map(ToolchainResult::Ok))
+        }),
+        ToolchainRequest::ImplementationVersion(implementation, version) => Box::new({
             debug!("Searching for {request} in {sources}");
-            let Some((source, interpreter)) =
-                python_interpreters(Some(version), Some(implementation), system, sources, cache)
-                    .find(|result| {
-                        match result {
-                            // Return the first critical error or matching interpreter
-                            Err(err) => should_stop_discovery(err),
-                            Ok((_source, interpreter)) => {
-                                version.matches_interpreter(interpreter)
-                                    && interpreter.implementation_name() == implementation.as_str()
-                            }
-                        }
-                    })
-                    .transpose()?
-            else {
-                // TODO(zanieb): Peek if there are any interpreters with the requested implementation
-                //               to improve the error message e.g. using `NoMatchingImplementation` instead
-                return Ok(ToolchainResult::Err(
-                    ToolchainNotFound::NoMatchingImplementationVersion(
-                        sources.clone(),
-                        *implementation,
-                        version.clone(),
-                    ),
-                ));
-            };
-            Toolchain {
-                source,
-                interpreter,
-            }
+            python_interpreters(Some(version), Some(implementation), system, sources, cache)
+                .filter(|result| match result {
+                    Err(_) => true,
+                    Ok((_source, interpreter)) => {
+                        version.matches_interpreter(interpreter)
+                            && interpreter.implementation_name() == implementation.as_str()
+                    }
+                })
+                .map(|result| result.map(Toolchain::from_tuple).map(ToolchainResult::Ok))
+        }),
+    }
+}
+
+/// Find a toolchain that satisfies the given request.
+///
+/// If an error is encountered while locating or inspecting a candidate toolchain,
+/// the error will raised instead of attempting further candidates.
+pub(crate) fn find_toolchain(
+    request: &ToolchainRequest,
+    system: SystemPython,
+    sources: &ToolchainSources,
+    cache: &Cache,
+) -> Result<ToolchainResult, Error> {
+    let mut toolchains = find_toolchains(request, system, sources, cache);
+    if let Some(result) = toolchains.find(|result| {
+        match result {
+            // Return the first critical discovery error or toolchain result
+            Err(err) => should_stop_discovery(err),
+            Ok(_) => true,
```

Comment:
Maybe... `result.map_err(|e| should_stop_discovery(&e)).err().unwrap_or(true)`? I think that works, but it's a little subtle because `err()` is turning the `Result<Foo, bool>` into `Option<bool>`. Could also be written `result.err().map_or(true, |e| should_stop_discovery(&e))`.

Reply:
Opting for `result.as_ref().err().map_or(true, should_stop_discovery)`